pkg/backend/filestate/lock.go

```diff
@@ -0,0 +1,157 @@
+		return nil, err
+	}
+	return &lockContent{
+		Pid:       os.Getpid(),
+		Username:  u.Username,
+		Hostname:  hostname,
+		Timestamp: time.Now(),
+	}, nil
+}
+
+func (l *lockContent) String() string {
+	return fmt.Sprintf("%v@%v (pid %v)", l.Username, l.Hostname, l.Pid)
+}
+
+func (b *localBackend) checkForLock(ctxt context.Context, stackRef backend.StackReference) error {
+	allFiles, err := listBucket(b.bucket, b.LockDir())
+	if err != nil {
+		return err
+	}
+
+	var lockKeys []string
+	for i := len(allFiles) - 1; i >= 0; i-- {
+		if allFiles[i].IsDir {
+			continue
+		}
+		fileKey := allFiles[i].Key
+		if strings.HasPrefix(fileKey, b.lockPrefix(stackRef.Name())) {
+			lockKeys = append(lockKeys, fileKey)
+		}
+	}
+
+	if len(lockKeys) > 0 {
+		errorString := fmt.Sprintf("the stack is current locked by %v lock(s). Either wait for the other "+
+			"process(es) to end or manually delete the lock file(s).", len(lockKeys))
+
+		for _, lock := range lockKeys {
+			content, err := b.bucket.ReadAll(ctxt, lock)
+			if err != nil {
+				return err
+			}
+			l := &lockContent{}
+			err = json.Unmarshal(content, &l)
+			if err != nil {
+				return err
+			}
+
+			// this is kind of weird but necessary because url is a string and not a url.URL
+			url := b.url + filepath.Join("/", lock)
+			errorString = errorString + fmt.Sprintf("\n  %v: created by %v", url, l.String())
+		}
+
+		return fmt.Errorf(errorString)
+	}
+	return nil
+}
+
+func (b *localBackend) checkForLockRace(stackRef backend.StackReference) error {
+	// Check the locks to make sure ONLY our lock exists. If we find multiple locks then we know
+	// we had a race condition and we should clean up and abort
+	allFiles, err := listBucket(b.bucket, b.LockDir())
+	if err != nil {
+		b.Unlock(stackRef)
+		return err
+	}
+
+	for i := len(allFiles) - 1; i >= 0; i-- {
+		file := allFiles[i].Key
+		if strings.HasPrefix(file, b.lockPrefix(stackRef.Name())) && file != b.lockPath(stackRef.Name()) {
+			b.Unlock(stackRef)
+			return fmt.Errorf("another Pulumi execution grabbed the lock, please wait" +
+				"for the other Pulumi process to complete and try again")
+		}
+	}
+	return nil
+}
+
+func (b *localBackend) Lock(stackRef backend.StackReference) error {
+	ctxt := context.TODO()
+
+	err := b.checkForLock(ctxt, stackRef)
+	if err != nil {
+		return err
+	}
+
+	lockContent, err := newLockContent()
+	if err != nil {
+		return err
+	}
+
+	content, err := json.Marshal(lockContent)
+	if err != nil {
+		return err
+	}
+
+	err = b.bucket.WriteAll(ctxt, b.lockPath(stackRef.Name()), content, nil)
+	if err != nil {
+		return err
+	}
+
+	err = b.checkForLockRace(stackRef)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *localBackend) Unlock(stackRef backend.StackReference) {
+	err := b.bucket.Delete(context.TODO(), b.lockPath(stackRef.Name()))
+	if err != nil {
+		logging.Errorf("there was a problem deleting the lock at %v: %v",
+			filepath.Join(b.url, b.lockPath(stackRef.Name())), err)
```

Comment:
should this not return an error as well.  basically, at this point, things are in a very bad state, and it looks like manual cleanup is necessary.  We may have to even explicitly state that that's the case.